design/object-store-user.md

```diff
@@ -0,0 +1,158 @@
+### Background
+This proposal is a necessary part of another enhancement to [dynamically provision S3 buckets](https://github.com/rook/rook/issues/2480). 
+
+With the merge of [2076](https://github.com/rook/rook/pull/2076) Rook now supports an object user CRD and its generated secret containing the AWS key pairs. To create a new object store user, a developer creates a simple _CephObjectStoreUser_ yaml file defining the object store and the user’s display name. After applying this file a CephObjectStoreUser CRD is created in the object store’s namespace, and the S3 object user is created.
+
+Some issues with this approach are:
+* the developer needs to know the name and namespace of the object store service, both to create the object store user yaml file, and to know, in advance, the name of the generated secret[1]. 
+* the object store user needs to reside in the same namespace as the target object store.
+
+<sup>[1]  It is desirable that the app developer be able to produce the pod yaml before creation of the object store user and its generated secret, and, thus, the secret’s name must be predictable since it is consumed by the pod. Today, the secret name consists of “rook-ceph-object-user-” + _objectStore’s metadata.Name_ + _CephObjectStoreUser’s metadata.Name_, and is assumed to live in the object store’s namespace. This enhancement proposes to remove the object store name from the secret’s name. Since object store user names must be unique within a namespace, a secret name derived from the CephObjectStoreUser’s `metadata.Name` is also unique, thus the object store name is not required. This supports an advantage of using a storage class (see below), namely, that the developer does not need to know the name (or namespace) of the object store.</sup>
+
+## Proposed Enhancement
+1. Allow a storage class, which references an object store service, to be defined in the object user API, but continue to support the existing `store` property.
+1. Remove the restriction that a CephObjectStoreUser be located in the same namespace as the ObjectStore.
+1. Change the ownerReference of the generated key-pair Secret to reference the CephObjectStoreUser rather than the ceph-cluster.
+1. Change the name of the generated secret per [1] above. Note: this change is **not** backwards compatible.
+
+#### Storage Class
+A [storage class](https://kubernetes.io/docs/concepts/storage/storage-classes/) is a global Kubernetes resource with the ability to pass parameters to a provisioner. Storage classes further abstract the underlying storage based on admin policies. Storage class visibility can be controlled by RBAC rules. A simple storage class for a rook-ceph object store might look like this:
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: rook-ceph-object-store
+provisioner: kubernetes.io/rook-ceph-operator
```

Comment:
does this provisioner field get used for anything?  or is it ignored, since consumers of this storage class will simply look up the object store name and namespace from the parameters directly?

Reply:
For the rook user `provisioner` is ignored -- it's the `objectStore` value that matters. However, for the bucket provisioner the storage class's provisioner matter. The bucket provisioner may do a `switch` based on the storage class provisioner.